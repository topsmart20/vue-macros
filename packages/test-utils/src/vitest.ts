import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
