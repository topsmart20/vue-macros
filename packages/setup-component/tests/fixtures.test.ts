import { resolve } from 'node:path'
import { describe } from 'vitest'
import {
  RollupEsbuildPlugin,
  RollupRemoveVueFilePathPlugin,
  RollupVue,
  RollupVueJsx,
  rollupBuild,
  testFixtures,
} from '@vue-macros/test-utils'
import VueSetupComponent from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    'tests/fixtures/*.{vue,[jt]s?(x)}',
    (args, id) =>
      rollupBuild(id, [
        VueSetupComponent(),
        RollupVue(),
        RollupVueJsx(),
        RollupRemoveVueFilePathPlugin(),
        RollupEsbuildPlugin({
          target: 'esnext',
        }),
      ]),
    {
      cwd: resolve(__dirname, '..'),
      promise: true,
    }
  )
})
