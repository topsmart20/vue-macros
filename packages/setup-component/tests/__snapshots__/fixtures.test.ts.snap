// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/SetupFC.tsx 1`] = `
"import { defineComponent, ref, createVNode, createTextVNode } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  __name: \\"SetupFC.tsx-setup-component-0\\",
  setup(__props) {
    const {
      App
    } = __MACROS_ctx();
    const count = ref(0);
    defineRender(() => createVNode(\\"div\\", null, [createVNode(\\"p\\", null, [createTextVNode(\\"hi, this is \\"), props.name]), createVNode(\\"p\\", null, [count.value]), createVNode(\\"button\\", {
      \\"onClick\\": () => count.value++
    }, [createTextVNode(\\"inc\\")])]));
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const App = __MACROS_setupComponent_0(() => ({
  App
}));

export { App };
"
`;

exports[`fixtures > tests/fixtures/async.tsx 1`] = `
"import { defineComponent, withAsyncContext } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  __name: \\"async.tsx-setup-component-0\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  async setup(__props) {
    let __temp, __restore;
    const {
      App
    } = __MACROS_ctx();
    ;
    [__temp, __restore] = withAsyncContext(() => globalThis.foo()), await __temp, __restore();
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const App = __MACROS_setupComponent_0(() => ({
  App
}));

export { App };
"
`;

exports[`fixtures > tests/fixtures/basic.ts 1`] = `
"import { defineComponent, h } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  ...{
    name: \\"App\\"
  },
  __name: \\"basic.ts-setup-component-0\\",
  props: {
    foo: { type: String, required: true }
  },
  emits: [\\"change\\"],
  setup(__props) {
    const { App } = __MACROS_ctx();
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const App = __MACROS_setupComponent_0(() => ({ App }));
h(App);

export { App };
"
`;

exports[`fixtures > tests/fixtures/context.tsx 1`] = `
"import { defineComponent } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main$1 = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  __name: \\"context.tsx-setup-component-1\\",
  setup(__props) {
    const {
      qux,
      quux,
      foo,
      bar,
      App,
      baz
    } = __MACROS_ctx();
    console.log(foo, bar, baz, qux, quux);
    return () => {
    };
  }
});

var __MACROS_setupComponent_1 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main$1(ctx), [__FILE__]);

const bar$1 = \\"BAR\\";
var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  __name: \\"context.tsx-setup-component-0\\",
  setup(__props) {
    const {
      foo,
      App,
      baz
    } = __MACROS_ctx();
    console.log(foo, bar$1, baz, App);
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const foo = \\"foo\\";
function bar() {
}
const App = __MACROS_setupComponent_0(() => ({
  foo,
  bar,
  App,
  baz
}));
{
  const qux = \\"qux\\";
  __MACROS_setupComponent_1(() => ({
    qux,
    quux,
    foo,
    bar,
    App,
    baz
  }));
  const quux = \\"quux\\";
}
var baz = \\"baz\\";

export { App };
"
`;

exports[`fixtures > tests/fixtures/render.tsx 1`] = `
"import { defineComponent, createVNode } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  __name: \\"render.tsx-setup-component-0\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  setup(__props) {
    const {
      App
    } = __MACROS_ctx();
    defineRender(() => createVNode(\\"div\\", null, null));
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const App = __MACROS_setupComponent_0(() => ({
  App
}));

export { App };
"
`;

exports[`fixtures > tests/fixtures/return-render.tsx 1`] = `
"import { defineComponent, computed, createVNode } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  __name: \\"return-render.tsx-setup-component-0\\",
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  setup(__props) {
    const {
      App
    } = __MACROS_ctx();
    const sum = computed(() => 1 + 2);
    defineRender(() => createVNode(\\"div\\", null, [sum.value]));
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const App = __MACROS_setupComponent_0(() => ({
  App
}));

export { App };
"
`;

exports[`fixtures > tests/fixtures/type-annotation.ts 1`] = `
"import { defineComponent, h } from 'vue';
import _export_sfc from '/plugin-vue/export-helper';

var _sfc_main = (__MACROS_ctx) => /* @__PURE__ */ defineComponent({
  ...{
    name: \\"App\\"
  },
  __name: \\"type-annotation.ts-setup-component-0\\",
  props: {
    foo: { type: String, required: true }
  },
  emits: [\\"change\\"],
  setup(__props) {
    const { App } = __MACROS_ctx();
    return () => {
    };
  }
});

var __MACROS_setupComponent_0 = (ctx) => /* @__PURE__ */ _export_sfc(_sfc_main(ctx), [__FILE__]);

const App = __MACROS_setupComponent_0(() => ({ App }));
h(App);

export { App };
"
`;
